### PR TITLE
apply: symlink in systemd files

### LINKED ISSUE #25: apply: propagate systemd units to runtime dir
Archives may contain systemd units which need to be propagate into the system. To make this ephemeral, we'd use the runtime systemd directory `/run/systemd/system/`. Symlinking is preferred (but not sure if it works), otherwise copying to there is fine as well.

### PR DESCRIPTION
**Pull request by squeed**
Fixes #25

### CODE REVIEW

File: pkg/torcx/types.go [hunk 1]
```diff
@@ -33,6 +33,8 @@ const (
 	DefaultTagRef = "com.coreos.cl"
 	// DEFAULT_PROFILE is the default profile name used
 	DEFAULT_PROFILE_NAME = "vendor"
+	// SYSTEMD_DIR is the directoy where we'll install all systemd files
+	SYSTEMD_DIR = "/run/systemd"
```
Comment (lucab) [inline pkg/torcx/types.go]: I'm confused, shouldn't this be `/run/systemd/transient/`?

Comment (squeed) [inline pkg/torcx/types.go]: Nah.

Comment (lucab) [inline pkg/torcx/types.go]: You are totally right, and it looks like so far I always confused `transient` with `runtime`. I've updated the original ticket.

**Comment by lucab** [comment]
It looks fine as a first step. It would probably need heavy changes once we settle on https://github.com/coreos/torcx/issues/29. Mind rebasing on current master to make travis happy?

File: pkg/torcx/types.go (hunk 1 above)
Comment (squeed) [inline pkg/torcx/types.go]: (duplicating OOB conversation)
For the record, `/run/systemd` is essentially a duplicate of `/usr/lib/systemd/...` but has precedence. `/run/systemd/transient` is only used for transient units created by the d-bus API or `systemd-run`. However, we might need to do things like install network files.

**Comment by squeed** [comment]
Rebased!

File: pkg/torcx/propagate.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/pkg/errors"
+)
+
+const ()
+
+// propagateUnits installs all systemd unit and unit-like files as transient
+// units in /run/systemd.
+// Units are taken from /usr/lib/systemd
+func propagateUnits(applyCfg *ApplyConfig, imageRoot string) error {
+
```
Comment (crawford) [inline pkg/torcx/propagate.go]: newline

File: pkg/torcx/propagate.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+			return err
+		}
+
+		// Copy src to dest
+		// * If src is dir, mkdir if not exists
+		// * If src is symlink, duplicate link contents exactly
+		// * If src is file, symlink
+		if srcStat.IsDir() {
+			if os.IsNotExist(err) {
+				return os.Mkdir(dstPath, 0755)
```
Comment (crawford) [inline pkg/torcx/propagate.go]: Is there any reason to create empty directories? Why not use MkdirAll when placing files instead?

File: pkg/torcx/propagate.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		} else if srcStat.Mode()&os.ModeSymlink > 0 { // Duplicate symlink
+			if !os.IsNotExist(dstStatErr) {
+				return fmt.Errorf("Cannot copy unit file %s to %s - source exists", srcPath, dstPath)
+			}
+			linkDest, err := os.Readlink(srcPath)
+			if err != nil {
+				return errors.Wrapf(err, "could not read link %s", srcPath)
+			}
+			// TODO(cdc) consider some link sanity checking
+			return os.Symlink(linkDest, dstPath)
```
Comment (crawford) [inline pkg/torcx/propagate.go]: This will only work for relative symlinks, right?

**Pull request merged**